### PR TITLE
Major updates to Contribution Guide

### PR DESCRIPTION
**Pull request by adamjstewart**
This PR involves several major updates to the Contribution Guide:

- Point to our Contribution Guide from the main Spack README
- Add documentation on Travis CI and how to run unit, flake8, and documentation tests locally
- Updates to the Git Workflows (see below)

@davydden Can you take a look at this since you wrote the original Contribution Guide?

I decided to put the instructions on how to run unit tests and flake8 tests above your git workflows in the hopes that people actually run these tests before submitting a PR. I also re-organized your git workflows a bit. I feel like branching and rebasing the "normal" way is probably easier for people new to Git than the cherry-pick methods, so I put them before the cherry-pick sections. I added a new section on rebasing with the "rebase" command.

@citibeth You would probably be interested in this too. Sorry it's taken me so long to document how to run tests locally. Let me know if there's anything you think is unclear or isn't documented.

P.S. It looks like the diff view is pretty much useless...

### CODE REVIEW

File: lib/spack/docs/contribution_guide.rst [hunk 1]
```diff
@@ -1,247 +1,514 @@
...
-and create a PR
+Now, we need to switch to the branch you submitted for your PR and rebase it
+on top of develop:
+
+.. code-block:: console
+
+   $ git checkout <descriptive_branch_name>
+   $ git rebase develop
+
+This will likely crash when it comes to the conflicting commit. Edit the file
```
Comment (davydden) [inline lib/spack/docs/contribution_guide.rst]: maybe reword `crash` into something less scary, along the lines of `git witll ask you to resolve conflicts`.

File: lib/spack/docs/contribution_guide.rst [hunk 2]
```diff
@@ -1,247 +1,514 @@
...
+that it says can't be merged automatically and resolve the conflict. Then, run:
```
Comment (davydden) [inline lib/spack/docs/contribution_guide.rst]: maybe mention that you can also run `git mergetool` to use GUI (such as `kdiff3`) to resolve the conflicts.

File: lib/spack/docs/contribution_guide.rst [hunk 3]
```diff
@@ -1,247 +1,514 @@
...
-First approach is as follows.
-You checkout your local develop branch, which for the purpose of this guide
-will be called ``develop_modified``:
+Unit tests ensure that core Spack features like fetching or spec resolution are
+working as expected. If your PR only adds new packages or modifies existing ones,
+there's very little chance that your changes could cause the unit tests to fail.
+However, if you make changes to Spack's core libraries, you should run the unit
+tests to make sure you didn't break anything.
+
+Since the unit tests test things like fetching from VCS repos, they require
```
Comment (adamjstewart) [inline lib/spack/docs/contribution_guide.rst]: Done.

File: lib/spack/docs/contribution_guide.rst [hunk 4]
```diff
@@ -1,247 +1,514 @@
...
 
-  $ git checkout develop_modified
-  $ git merge <descriptive_branch_name>
+   $ git add <files_to_be_part_of_the_commit>
+   $ git commit --message <descriptive_message_of_this_particular_commit>
 
-The net result is similar to the first approach with a minor difference that
-you would also merge upstream develop into you modified version in the last
-step. Should this not be desirable, you have to follow the first approach.
+Now, you can push it to your remote fork and create a PR:
```
Comment (adamjstewart) [inline lib/spack/docs/contribution_guide.rst]: Done.

File: lib/spack/docs/contribution_guide.rst (hunk 1 above)
Comment (adamjstewart) [inline lib/spack/docs/contribution_guide.rst]: Done.

File: lib/spack/docs/contribution_guide.rst (hunk 2 above)
Comment (adamjstewart) [inline lib/spack/docs/contribution_guide.rst]: I've never used `git mergetool` before so you'd have to give me a blurb to use. But then we would have to explain how to set up `git mergetool`. I don't want to make this too complicated. We just need to give the bare minimum explanation of how to work with Git for new users.

Comment (davydden) [inline lib/spack/docs/contribution_guide.rst]: i find this very helpful once you are used to specific GUI of the tool. 

You can set mergetool by `git config --global merge.tool kdiff3` or use it only for a single merge `git mergetool --tool=kdiff3`. To check the list of tools `git mergetool --tool-help`. Here is some GUI docu for kdiff3 http://kdiff3.sourceforge.net/doc/merging.html 

Feel free to use some of none of this at all. :smile:

Comment (davydden) [inline lib/spack/docs/contribution_guide.rst]: it is certainly possible:
```
$ git mergetool --tool-help
'git mergetool --tool=<tool>' may be set to one of the following:
		emerge
		opendiff
		vimdiff
		vimdiff2
		vimdiff3

	user-defined:
		kdiff3NoAuto.cmd /Applications/kdiff3.app/Contents/MacOS/kdiff3 --L1 "$MERGED (Base)" --L2 "$MERGED (Local)" --L3 "$MERGED (Remote)" -o "$MERGED" "$BASE" "$LOCAL" "$REMOTE"

The following tools are valid, but not currently available:
		araxis
		bc
		bc3
		codecompare
		deltawalker
		diffmerge
		diffuse
		ecmerge
		gvimdiff
		gvimdiff2
		gvimdiff3
		kdiff3
		meld
		p4merge
		tkdiff
		tortoisemerge
		winmerge
		xxdiff
```

**Comment by tgamblin** [comment]
Is this ready to merge?

**Comment by adamjstewart** [comment]
Yep, should be good to go. Fun fact: if you click on the "View" button for a `.rst` or `.md` file, GitHub will actually render the page for you so you can preview the result. It won't do inter-file links or notes, but it helps to catch little bugs like the ones I fixed in the last commit. Somehow I had never noticed this functionality before.

File: lib/spack/docs/contribution_guide.rst [hunk 5]
```diff
@@ -1,247 +1,514 @@
...
+Flake8 Tests
+^^^^^^^^^^^^
+
+Spack uses `Flake8 <http://flake8.pycqa.org/en/latest/>`_ to test for
+`PEP 8 <https://www.python.org/dev/peps/pep-0008/>`_ conformance. PEP 8 is
+a series of style guides for Python that provide suggestions for everything
+from variable naming to indentation. In order to limit the number of PRs that
+were mostly style changes, we decided to enforce PEP 8 conformance. If your
+PR does not comply with PEP 8, it will not be accepted.
+
```
Comment (citibeth) [inline lib/spack/docs/contribution_guide.rst]: Reword positively to make Spack devs sound nicer: "Your PR needs to comply with PEP 8 to be accepted."

File: lib/spack/docs/contribution_guide.rst [hunk 6]
```diff
@@ -1,247 +1,514 @@
...
+   var/spack/repos/builtin/packages/netcdf/package.py:26: [F401] 'os' imported but unused
+   var/spack/repos/builtin/packages/netcdf/package.py:61: [E303] too many blank lines (2)
+   var/spack/repos/builtin/packages/netcdf/package.py:106: [E501] line too long (92 > 79 characters)
+   Flake8 found errors.
+
+Most of the error messages are straightforward, but if you don't understand what
+they mean, just ask questions about them when you submit your PR. The line numbers
+will change if you add or delete lines, so simply run ``run-flake8-tests`` again
+to update them.
+
```
Comment (citibeth) [inline lib/spack/docs/contribution_guide.rst]: Add something like:
```
.. tip::
   Try fixing flake8 errors in reverse order.  Eliminates the need for multiple runs of `flake8` just to re-compute line numbers; and it makes it much easier to fix errors directly off of the Travis output.
```
